### PR TITLE
Movement experiments

### PR DESCRIPTION
**Pull request by Cheeseness**
Merge of initial work on basic movement and camera controls.

### CODE REVIEW

File: src/Player/Camera/Camera.gd [hunk 1]
```diff
@@ -1,7 +1,86 @@
-extends Camera
+extends Spatial
 
 var _target: Spatial
+export var camera_rotation_speed: = Vector2(2.5, 2.5)
```
Comment (NathanLovato) [inline src/Player/Camera/Camera.gd]: Don't repeat the class name in properties

File: src/Player/Camera/Camera.gd [hunk 2]
```diff
@@ -1,7 +1,86 @@
...
+export var yinvert_multiplier: = -1
```
Comment (NathanLovato) [inline src/Player/Camera/Camera.gd]: Is this necessary? Isn't a multiplier that inverts a value always -1?

File: src/Player/Camera/Camera.gd [hunk 3]
```diff
@@ -1,7 +1,86 @@
...
+export var camera_backwards_deadzone := 0.3
+var _camera_rotating: = false
+
+onready var initial_camera_pos = $Camera.get_translation()
```
Comment (NathanLovato) [inline src/Player/Camera/Camera.gd]: No need for camera here too, and use `position` as in Godot

File: src/Player/Camera/Camera.gd [hunk 4]
```diff
@@ -1,7 +1,86 @@
...
 
 func _ready() -> void:
 	_target = owner
 	set_as_toplevel(true)
+	occlusion_ray.set_cast_to(initial_camera_pos)
+
+
+func physics_process(delta: float, move_direction: Vector3) -> void:
+	#TODO: Remove move direction visual aid - or turn it into a gizmo?
+	$MoveDirectionRef.set_translation(move_direction)
```
Comment (NathanLovato) [inline src/Player/Camera/Camera.gd]: All node dependencies should be `onready var ...` at the top of the variable definitions

File: src/Player/Camera/Camera.gd [hunk 5]
```diff
@@ -1,7 +1,86 @@
...
 var _target: Spatial
+export var camera_rotation_speed: = Vector2(2.5, 2.5)
+export var yinvert_multiplier: = -1
+export var camera_backwards_deadzone := 0.3
+var _camera_rotating: = false
+
+onready var initial_camera_pos = $Camera.get_translation()
+onready var initial_anchor_pos = get_translation()
+onready var rotation_delay: Timer = $RotateDelay
+onready var occlusion_ray: RayCast = $OcclusionRay
```
Comment (NathanLovato) [inline src/Player/Camera/Camera.gd]: Node dependencies, i.e. `onready var ... = $...` should come first

File: src/Player/Camera/Camera.gd [hunk 6]
```diff
@@ -1,7 +1,86 @@
...
+
+	if _camera_rotating == false:
+		rotation_delay.stop()
+
+	#If there is a body between the camera and the player, move the camera closer
+	if occlusion_ray.is_colliding():
+		var offset_pos = $Camera.get_global_transform()
+		if offset_pos.origin != occlusion_ray.get_collision_point():
+			offset_pos.origin = occlusion_ray.get_collision_point()
+			$Camera.set_global_transform(offset_pos)
```
Comment (NathanLovato) [inline src/Player/Camera/Camera.gd]: Same, `Camera` should be stored in an `onready var`

**Comment by NathanLovato** [comment]
I left some notes about code style as it's the first review. I can make the changes for you in a new commit if you want.

File: src/Player/Camera/Camera.gd (hunk 2 above)
Comment (Cheeseness) [reply src/Player/Camera/Camera.gd]: Different people have different inversion preferences. Exposing a value that developers can set in the editor to control default behaviour feels like the right way to go.

Making it use an enum seems to provide a dropdown in the editor, which is a bit clearer. I'll switch it up to use that.

**Comment by Cheeseness** [comment]
Thanks! I can make these changes tonight.

**Comment by NathanLovato** [comment]
Thanks, just merged :)

**Pull request merged**